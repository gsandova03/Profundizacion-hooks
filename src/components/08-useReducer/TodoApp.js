import React, { useReducer } from 'react'

import './styles/styles.css'
import { todoReducer } from './todoReducer'

const initialState = [
  {
    id: new Date().getTime(),
    todo: 'Aprender React',
    done: false
  },
  {
    id: new Date().getTime(),
    todo: 'Aprender Node',
    done: false
  }
]

export const TodoApp = () => {

  const [ todos, dispatch ] = useReducer( todoReducer, initialState )

  const handleAddSubmit = (e) => {
    e.preventDefault();

    const newTodo = {
      id: new Date().getTime(),
      todo: 'Aprender Express',
      done: false
    };

    const action = {
      type: "addTodo",
      payload : newTodo
    };

    dispatch( action );
  }
  
  return (
    <div>
      <h1>Tareas sin completar: ( { todos.length } ) </h1>
      {/* <h1>Tareas terminadas: ( { todos.length } ) </h1> */}
      <hr/>

      <div className="row">
        <div className="col-7">
          <h4 className="text-center">Todos</h4>
          <hr/>
          <ul className="list-group list-group-flush">
            {
              todos.map( ( todo,i ) => (
                <li key={ todo.id } className="list-group-item">
                  <p className="text-center"> { i + 1 }. { todo.todo } </p>
                  <button className="btn btn-danger"> Eliminar </button>
                </li>
              ))
            }
          </ul>
        </div>

        <div className="col-5" onSubmit={ handleAddSubmit }>
            <h4 className="text-center">Aregar Todo</h4>
            <hr/>
            <form>
              <input 
                type="text" 
                name="description" 
                className="form-control"
                placeholder="tarea...." 
              />
              <button 
                type="submit" 
                className="btn btn-primary"
              >
                Agregar
              </button>
            </form>
        </div>
      </div>

    </div>
  )
}
