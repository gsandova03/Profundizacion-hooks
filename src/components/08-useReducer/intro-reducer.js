export const initialState = [{
  id: 1,
  todo: 'Comprar Pan',
  done: false
}]

const todoReducer = ( state = initialState , action ) => {

  switch (action?.type) {
    case "addTodo":
      return [ ...state, action.payload ]
  
    default:
      return state
  }
}

let todos = todoReducer();

let newTodo = {
  id: 2,
  todo: 'Sacar al perro',
  done: true
}

const action = {
  type: 'addTodo',
  payload: newTodo
}

todos = todoReducer( todos, action )

console.log( todos )