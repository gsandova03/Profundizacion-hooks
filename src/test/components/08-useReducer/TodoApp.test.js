import React from 'react'
import { shallow, mount } from "enzyme"
import { TodoApp } from '../../../components/08-useReducer/TodoApp'
import { demoTodo } from '../../fixtures/demoTodo'
import { act } from '@testing-library/react'

describe('Pruebas en el componente <TodoApp />', () => {
  
  const wrapper = shallow( <TodoApp/> )
  //Para hacer validaciones del llamdo al localStorage jest nos permite utilizar la funcion Storage
  Storage.prototype.setItem = jest.fn()
  
  test('debe de mostrarse correctamente', () => {  
    expect( wrapper ).toMatchSnapshot()
  });
  
  test('debe de agregar un Todo', () => {
    
    //obtenemos mas informacion del componente y de sus hijos con el mount, ademas del contexto donde esta corriendo
    const wrapper = mount( <TodoApp/> )

    act(()=>{
      wrapper.find('TodoAdd').prop('addTodo')( demoTodo[0] );
      wrapper.find('TodoAdd').prop('addTodo')( demoTodo[1] );
    })

    expect( wrapper.find('h1').text().trim() ).toBe( 'Tareas sin completar: ( 2 )' );
    //Evaluacion del llamdo al localStorage
    expect( localStorage.setItem ).toHaveBeenCalledTimes(2);
  })
  
  test('debe de eliminar un todo', () => {
    
    wrapper.find('TodoAdd').prop('addTodo')( demoTodo[0] );
    wrapper.find('TodoList').prop('handleDelete')( demoTodo[0].id );

    expect( wrapper.find('h1').text().trim() ).toBe( 'Tareas sin completar: ( 0 )' );
  })
  
})
