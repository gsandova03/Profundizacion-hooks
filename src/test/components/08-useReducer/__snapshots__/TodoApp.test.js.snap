// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <TodoApp /> debe de mostrarse correctamente 1`] = `
<div>
  <h1>
    Tareas sin completar: ( 
    0
     ) 
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <h4
        className="text-center"
      >
        Todos
      </h4>
      <hr />
      <TodoList
        handleComplete={[Function]}
        handleDelete={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <TodoAdd
        addTodo={[Function]}
      />
    </div>
  </div>
</div>
`;
